source/unified-test-format/unified-test-format.rst

```diff
@@ -0,0 +1,1461 @@
+    be created. The YAML file SHOULD use an `alias node`_ for a database
+    entity's ``id`` field (e.g. ``database: *database0``).
+
+  - ``bucketOptions``: Optional document. Additional options used to construct
+    the bucket object. Supported options are defined in the
+    `GridFS <../source/gridfs/gridfs-spec.rst#configurable-gridfsbucket-class>`__
+    specification. The ``readConcern``, ``readPreference``, and ``writeConcern``
+    options use the same structure as defined in `Common Options`_.
+
+
+collectionData
+~~~~~~~~~~~~~~
+
+List of documents that should correspond to the contents of a collection. This
+structure is used by both `initialData`_ and `test.outcome <test_outcome_>`_,
+which insert and read documents, respectively. Both of those directives may
+operate on the collection(s) under test, they do not share the same Collection
+and MongoClient object(s) as test `operations <operation_>`_. The structure of
+this document is as follows:
+
+- ``collectionName``: Optional string. Collection name (not an `entity`_).
+  Defaults to the name of the collection under test (see: `collectionName`_).
+
+- ``databaseName``: Optional string. Database name (not an `entity`_). Defaults
+  to the name of the database under test (see: `databaseName`_).
+
+- ``documents``: Required array of documents. List of documents corresponding to
+  the contents of the collection. This list may be empty.
+
+
+test
+~~~~
+
+Test case consisting of a sequence of operations to be executed. The test may
+optionally include configuration directives and event/outcome assertions. The
+structure of each document is as follows:
+
+- ``description``: Required string. The name of the test.
+
+.. _test_runOn:
+
+- ``runOn``: Optional array of documents. List of server version and/or topology
+  requirements for which the tests in this file can be run. If specified, these
+  requirements override any top-level requirements in `runOn`_. Test runners
+  MUST skip a test if its requirements are not met.
+
+  If set, the array should contain at least one document. The structure of each
+  document is defined in `runOnRequirement`_.
+
+- ``skipReason``: Optional string. If set, the test will be skipped. The string
+  should explain the reason for skipping the test (e.g. JIRA ticket).
+
+.. _test_operations:
+
+- ``operations``: Required array of documents. List of operations to be executed
+  for the test case.
+
+  The array should contain at least one document. The structure of each
+  document is defined in `operation`_.
+
+.. _test_expectedEvents:
+
+- ``expectedEvents``: Optional array of documents. List of events, which are
+  expected to be observed in this order by running the operations.
+
+  The array should contain at least one document. The structure of each
+  document is defined in `expectedEvent`_.
+
+  **TODO**: Determine if an empty array should test that no events are observed.
```

Comment:
An empty array now corresponds with an assertion that no events are observed.

To my second point, `test.expectedEvents` is now a document that associates a particular client with a list of expected events.

We presently only observe APM, so the issue of seeing SDAM events interspersed between APM events isn't an issue. I think the best approach would be to allow client entities to configure exactly which events can be collected. Most tests would just enable APM. Let me know if that makes sense and I'll add an `observeEvents` (takes an array of strings) key to the client entity definition.

Reply:
Will explicitly define `observeEvents` at the client entity level, and support "apm" out of the gate. We anticipate adding "sdam" later on.

We don't anticipate to ever need to consider the order of events across different clients, so `expectedEvents` can remain as is (list of events per client).

Reply:
I just realized that "APM" is a very generic term and more correctly refers to any kind of event listeners (e.g. command monitoring, SDAM). We seem to colloquially use it to refer to command monitoring but perhaps that's because it was the first.

Anywho, rather than enable types of event listener I think it's more straightforward and intuitive to specify a list of events that we want to observe, since those will already be defined in `expectedEvent` (we currently just have the three command monitoring events). Added benefit here is that tests won't have to completely enable _all_ command monitoring events, which will help with porting most specs that only care about CommandStartedEvent.

Reply:
Added client-level `observeEvents` based on the description above. See changelog.